fraud/service.go

```diff
@@ -1,57 +1,93 @@
 package fraud
 
 import (
+	"bytes"
 	"context"
 	"encoding/hex"
 	"errors"
 	"fmt"
 	"sync"
-	"time"
 
 	"github.com/ipfs/go-datastore"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
-)
+	"golang.org/x/sync/errgroup"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
 
-const (
-	// fetchHeaderTimeout duration of GetByHeight request to fetch an ExtendedHeader.
-	fetchHeaderTimeout = time.Minute * 2
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+	"github.com/celestiaorg/celestia-node/params"
 )
 
-// service is responsible for validating and propagating Fraud Proofs.
+// fraudRequests is amount of external requests that will be done in order to get fraud proofs from another peers.
+const fraudRequests = 5
+
+var fraudProtocolID = protocol.ID(fmt.Sprintf("/fraud/v0.0.1/%s", params.DefaultNetwork()))
+
+// ProofService is responsible for validating and propagating Fraud Proofs.
 // It implements the Service interface.
-type service struct {
+type ProofService struct {
 	topicsLk sync.RWMutex
 	topics   map[ProofType]*pubsub.Topic
 
 	storesLk sync.RWMutex
 	stores   map[ProofType]datastore.Datastore
 
 	pubsub *pubsub.PubSub
+	host   host.Host
 	getter headerFetcher
 	ds     datastore.Datastore
+
+	enabledSyncer bool
 }
 
-func NewService(p *pubsub.PubSub, getter headerFetcher, ds datastore.Datastore) Service {
-	return &service{
-		pubsub: p,
-		getter: getter,
-		topics: make(map[ProofType]*pubsub.Topic),
-		stores: make(map[ProofType]datastore.Datastore),
-		ds:     ds,
+func NewService(
+	p *pubsub.PubSub,
+	host host.Host,
+	getter headerFetcher,
+	ds datastore.Datastore,
+	enabledSyncer bool,
+) *ProofService {
+	return &ProofService{
+		pubsub:        p,
+		host:          host,
+		getter:        getter,
+		topics:        make(map[ProofType]*pubsub.Topic),
+		stores:        make(map[ProofType]datastore.Datastore),
+		ds:            ds,
+		enabledSyncer: enabledSyncer,
 	}
 }
 
-func (f *service) Subscribe(proofType ProofType) (_ Subscription, err error) {
+func (f *ProofService) RegisterTopics(proofTypes ...ProofType) error {
+	for _, proofType := range proofTypes {
+		t, err := join(f.pubsub, proofType, f.processIncoming)
+		if err != nil {
+			return err
+		}
+		f.topicsLk.Lock()
+		f.topics[proofType] = t
+		f.topicsLk.Unlock()
+	}
+	return nil
+}
+func (f *ProofService) Start(ctx context.Context) error {
+	f.host.SetStreamHandler(fraudProtocolID, f.handleFraudMessageRequest)
```

Comment:
Maybe this can be moved into a separate method on ProofService and can be `fx.Invoke` on bridge and full, and the fraud service lifecycle then should only be added for light node. Can remove this ugly `enableSyncer`

Reply:
Nah. Registering happens for each node type because each node type serves fps.